src/std/net/s3/api.ss

```diff
@@ -0,0 +1,242 @@
+  (lambda (self
+            (endpoint "s3.amazonaws.com")
+            (access-key (getenv "AWS_ACCESS_KEY_ID" #f))
+            (secret-key (getenv "AWS_SECRET_ACCESS_KEY" #f))
+            (region (getenv "AWS_DEFAULT_REGION" "us-east-1")))
+    (using (self :- s3-client)
+      (set! self.endpoint endpoint)
+      (set! self.access-key access-key)
+      (set! self.secret-key secret-key)
+      (set! self.region region))))
+
+; Retrieves buckets accessible to this client.
+(defmethod {list-buckets s3-client} ; => (list : bucket)
+  (lambda (self)
+    (using (self :- s3-client)
+           (let* ((req (s3-request/error self verb: 'GET))
+                  (xml (s3-parse-xml req))
+                  (buckets (sxml-find xml (sxml-e? 's3:Buckets) sxml-children))
+                  (names (map (chain <>
+                                     (sxml-select <> (sxml-e? 's3:Name))
+                                     (cadar <>)
+                                     (make-bucket self <> (s3-client-region self)))
+                              buckets)))
+           ; buckets is #f if none are returned
+           (request-close req)
+           names))))
+
+;; NOTE: all bucket operations need the correct region for the bucket or they will 400
+(defmethod {create-bucket! s3-client}
+  (lambda (self bucket)
+    (using (self :- s3-client)
+      (let (req (s3-request/error self verb: 'PUT bucket: bucket))
+        (request-close req)
+        (void)))))
+
+; Gets a bucket struct that can be used to fetch objects.
+(defmethod {get-bucket s3-client} ; => bucket
+  (lambda (self bucket-name)
+    (using (self :- s3-client)
+      (if (s3-client::bucket-exists? self bucket-name)
+        (make-bucket self bucket-name self.region)
+        #f))))
+
+; Delete a bucket by name
+(defmethod {delete-bucket! s3-client}
+  (lambda (self bucket)
+    (using (self :- s3-client)
+      (when (s3-client::bucket-exists? self bucket)
+        (let (req (s3-request/error self verb: 'DELETE bucket: bucket))
+          (request-close req)
+          (void))))))
+
+(defmethod {bucket-exists? s3-client}
+  (lambda (self bucket)
+    (using (self :- s3-client)
+           (let* ((bucket (if (bucket? bucket) (bucket-name bucket) bucket))
+                  (req {self.request verb: 'HEAD bucket: bucket})
+                  (code (request-status req)))
+             ; 200 and 404 are expected codes
+             ; we explicitly handle 404 so we get proper predicate
+             ; semantics and don't raise on what would otherwise be
+             ; #f condition.
+             (if (memv code [200 404]) 
+               (begin
+                 (request-close req)
+                 (= code 200))
+               (with-request-error req))))))
+
+(defmethod {bucket s3-client}
+  (lambda (self name)
+    (using (self self :- s3-client)
+      (if (s3-client::bucket-exists? self name)
+        (make-bucket self name (s3-client-region self))
+        #f))))
```

Comment:
again, interface wrap and inverted logic.

Reply:
Same deal here. I don't think this is inverted. I'll wrap it in the interface cast though